From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp90: Add the money, messages and time_get<unsigned short> facets in _Make*loc.

locale::_Locimp::_Makexloc, _Makewloc and _Makeushloc build the facets of the
categories asked for. msvcp80 and msvcp90 now add the messages, money_get,
money_put and moneypunct<false/true> facets of their character type, and
_Makeushloc adds time_get<unsigned short> in every version; before, a locale
made from a name had none of them. msvcp100 and later have no messages or
money facets yet, so their FIXME stays.

Building moneypunct<wchar_t> and <unsigned short> from a _Locinfo passed a
NULL conversion state to _Mbrtowc (patch 4538), which crashed in the "C"
locale; it now passes a zeroed state.

Gate: test/msvcp-makeloc-gate.sh (msvcp80 .. msvcp140, 64/32).
---
 dlls/msvcp90/locale.c | 87 +++++++++++++++++++++++++++++++++++++++++++++------
 1 file changed, 77 insertions(+), 10 deletions(-)

diff --git a/dlls/msvcp90/locale.c b/dlls/msvcp90/locale.c
index cc9c0b3..e17c915 100644
--- a/dlls/msvcp90/locale.c
+++ b/dlls/msvcp90/locale.c
@@ -1659,12 +1659,14 @@ static wchar_t *money_wcsdup_mb(const char *s, const _Locinfo *locinfo)
 {
     size_t len = strlen(s) + 1, i;
     wchar_t *ret = operator_new(len * sizeof(wchar_t));
+    _Mbstatet state;
     _Cvtvec cvt;
 
     _Locinfo__Getcvt(locinfo, &cvt);
+    memset(&state, 0, sizeof(state));
     for (i = 0; *s; i++)
     {
-        int r = _Mbrtowc(ret + i, s, MB_LEN_MAX, NULL, &cvt);
+        int r = _Mbrtowc(ret + i, s, MB_LEN_MAX, &state, &cvt);
         if (r <= 0) { ret[i] = (unsigned char)*s; r = 1; }
         s += r;
     }
@@ -15246,14 +15248,64 @@ locale__Locimp** __cdecl locale__Locimp__Clocptr_func(void)
     return &locale__Locimp__Clocptr;
 }
 
+/* the facet of loc (or the one its category makes), as use_facet gives it */
+static locale_facet* makeloc_use_facet(const locale *loc, locale_id *id,
+        void (__cdecl *getcat)(const locale_facet**, const locale*), locale_facet **cache)
+{
+    _Lockit lock;
+    const locale_facet *fac;
+
+    _Lockit_ctor_locktype(&lock, _LOCK_LOCALE);
+    fac = locale__Getfacet(loc, locale_id_operator_size_t(id));
+    if(!fac && *cache)
+        fac = *cache;
+    else if(!fac) {
+        getcat(&fac, loc);
+        *cache = (locale_facet*)fac;
+        call_locale_facet__Incref(*cache);
+        locale_facet_register(*cache);
+    }
+    _Lockit_dtor(&lock);
+    return (locale_facet*)fac;
+}
+
+/* adds the messages, money_get, money_put and both moneypunct facets of one
+ * character type to a _Locimp being made */
+#ifdef SG_MUTANT_MAKELOC_NO_MONEY
+#define MAKELOC_FACET(name, type, ...)
+#else
+#define MAKELOC_FACET(name, type, ...) \
+    if(cat & (1<<(name##__Getcat(NULL, NULL)-1))) { \
+        static locale_facet *cache; \
+        locale_facet *fac; \
+        if(loc) { \
+            fac = makeloc_use_facet(loc, &name##_id, (void *)name##__Getcat, &cache); \
+        }else { \
+            fac = operator_new(sizeof(type)); \
+            name##_ctor_locinfo((type*)fac, locinfo, 0 __VA_ARGS__); \
+        } \
+        locale__Locimp__Addfac(locimp, fac, locale_id_operator_size_t(&name##_id)); \
+    }
+#endif
+#if _MSVCP_VER <= 90
+#define MAKELOC_MONEY_FACETS(t) \
+    MAKELOC_FACET(messages_##t, messages) \
+    MAKELOC_FACET(money_get_##t, money_facet) \
+    MAKELOC_FACET(money_put_##t, money_facet) \
+    MAKELOC_FACET(moneypunct_##t##_local, mpunct, , FALSE) \
+    MAKELOC_FACET(moneypunct_##t##_intl, mpunct, , FALSE)
+#endif
+
 /* ?_Makeushloc@_Locimp@locale@std@@CAXABV_Locinfo@3@HPAV123@PBV23@@Z */
 /* ?_Makeushloc@_Locimp@locale@std@@CAXAEBV_Locinfo@3@HPEAV123@PEBV23@@Z */
-/* List of missing facets:
- * messages, money_get, money_put, moneypunct, moneypunct, time_get
- */
+/* msvcp100 and later have no messages, money_get, money_put or moneypunct facets */
 void __cdecl locale__Locimp__Makeushloc(const _Locinfo *locinfo, category cat, locale__Locimp *locimp, const locale *loc)
 {
+#if _MSVCP_VER <= 90
+    TRACE("(%p %d %p %p)\n", locinfo, cat, locimp, loc);
+#else
     FIXME("(%p %d %p %p) semi-stub\n", locinfo, cat, locimp, loc);
+#endif
 
     if(cat & (1<<(ctype_short__Getcat(NULL, NULL)-1))) {
         ctype_wchar *ctype;
@@ -15338,16 +15390,23 @@ void __cdecl locale__Locimp__Makeushloc(const _Locinfo *locinfo, category cat, l
         }
         locale__Locimp__Addfac(locimp, &codecvt->base.facet, locale_id_operator_size_t(&codecvt_short_id));
     }
+
+    MAKELOC_FACET(time_get_short, time_get_wchar)
+#if _MSVCP_VER <= 90
+    MAKELOC_MONEY_FACETS(short)
+#endif
 }
 
 /* ?_Makewloc@_Locimp@locale@std@@CAXABV_Locinfo@3@HPAV123@PBV23@@Z */
 /* ?_Makewloc@_Locimp@locale@std@@CAXAEBV_Locinfo@3@HPEAV123@PEBV23@@Z */
-/* List of missing facets:
- * messages, money_get, money_put, moneypunct, moneypunct
- */
+/* msvcp100 and later have no messages, money_get, money_put or moneypunct facets */
 void __cdecl locale__Locimp__Makewloc(const _Locinfo *locinfo, category cat, locale__Locimp *locimp, const locale *loc)
 {
+#if _MSVCP_VER <= 90
+    TRACE("(%p %d %p %p)\n", locinfo, cat, locimp, loc);
+#else
     FIXME("(%p %d %p %p) semi-stub\n", locinfo, cat, locimp, loc);
+#endif
 
     if(cat & (1<<(ctype_wchar__Getcat(NULL, NULL)-1))) {
         ctype_wchar *ctype;
@@ -15444,16 +15503,21 @@ void __cdecl locale__Locimp__Makewloc(const _Locinfo *locinfo, category cat, loc
         }
         locale__Locimp__Addfac(locimp, &codecvt->base.facet, locale_id_operator_size_t(&codecvt_wchar_id));
     }
+#if _MSVCP_VER <= 90
+    MAKELOC_MONEY_FACETS(wchar)
+#endif
 }
 
 /* ?_Makexloc@_Locimp@locale@std@@CAXABV_Locinfo@3@HPAV123@PBV23@@Z */
 /* ?_Makexloc@_Locimp@locale@std@@CAXAEBV_Locinfo@3@HPEAV123@PEBV23@@Z */
-/* List of missing facets:
- * messages, money_get, money_put, moneypunct, moneypunct
- */
+/* msvcp100 and later have no messages, money_get, money_put or moneypunct facets */
 void __cdecl locale__Locimp__Makexloc(const _Locinfo *locinfo, category cat, locale__Locimp *locimp, const locale *loc)
 {
+#if _MSVCP_VER <= 90
+    TRACE("(%p %d %p %p)\n", locinfo, cat, locimp, loc);
+#else
     FIXME("(%p %d %p %p) semi-stub\n", locinfo, cat, locimp, loc);
+#endif
 
     if(cat & (1<<(ctype_char__Getcat(NULL, NULL)-1))) {
         ctype_char *ctype;
@@ -15550,6 +15614,9 @@ void __cdecl locale__Locimp__Makexloc(const _Locinfo *locinfo, category cat, loc
         }
         locale__Locimp__Addfac(locimp, &codecvt->base.facet, locale_id_operator_size_t(&codecvt_char_id));
     }
+#if _MSVCP_VER <= 90
+    MAKELOC_MONEY_FACETS(char)
+#endif
 }
 
 /* ?_Makeloc@_Locimp@locale@std@@CAPAV123@ABV_Locinfo@3@HPAV123@PBV23@@Z */
